[sysvabi64] Add chapter on Thread Local Storage

smithp35 opened the pull request:
The thread local storage chapter contains:
* A description of Thread Local Storage based on addenda32
* The key design decisions of AArch64 TLS such as tls variant, tls dialect, TCB size.
* The ABI required code sequence for TLSDESC that must be emitted exactly, as GNU ld requires it to be.
* Sequences for the different code-models.
* Relaxations for GD->IE, GD->LE and IE->LE.
* Synchronization requirements for Lazy TLSDESC. With advice not to support it due to overhead of synchronization.

smithp35 committed:
[sysvabi64] Add chapter on Thread Local Storage

The thread local storage chapter contains:
* A description of Thread Local Storage based on addenda32
* The key design decisions of AArch64 TLS such as tls variant,
  tls dialect, TCB size.
* The ABI required code sequence for TLSDESC that must be emitted
  exactly, as GNU ld requires it to be.
* Sequences for the different code-models.
* Relaxations for GD->IE, GD->LE and IE->LE.
* Synchronization requirements for Lazy TLSDESC. With advice not
  to support it due to overhead of synchronization.

MaskRay commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+and ``PT_TLS`` as the program header with type PT_TLS. ``PAD`` must be
+the smallest positive integer that satisfies the following congruence:
+
+``TP + TCB + PAD ≡ PT_TLS.p_vaddr (modulo PT_TLS.p_align)``
```
`TP+TCB+PAD` on the left could be confusing, as TCB is placed before TP. Perhaps mention the requirement of TP first (= 0 (modulo p_align)), then describe `PAD` and this formula.

smithp35 replied:
I'll see if I can word it better. I've found it difficult to try and explain the formula intuitively.

MaskRay commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+Given that ``TP ≡ 0 (modulo PT_TLS.p_align)``. An expression
+for `PAD` is ``PAD = (PT_TLS.p_vaddr - TCB) mod PT_TLS.p_align``.
+
+A significant number of dynamic linkers use a different calculation
```
I believe that glibc Variant I handles `p_vaddr!=0 (mod p_align)` correctly. The bug (https://sourceware.org/bugzilla/show_bug.cgi?id=24606
) is for Variant II (x86 etc).

I have fixed FreeBSD rtld's Variant II in https://reviews.freebsd.org/D31538 . Its Variant I may or may not have the bug.

musl has been good since 1.1.23

Therefore, it's probably not "a significant number" but yeah `p_vaddr=0 (mod p_align)` is good for maximum compatibility

MaskRay commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+    add  xn, tp, :tprel_hi12:var, lsl #12 // R_AARCH64_TLSLE_ADD_TPREL_HI12 var
+    ldr  xn, [xn, #:tprel_lo12_nc:var]    // R_AARCH64_TLSLE_LDST64_TPREL_LO12_NC var
+
+Static link time TLS Relaxations
```
Perhaps call this Optimization to be consistent with x86/ppc and "Relocation optimization" (ADRP) and leave the term "relocation relaxation"  for RISC-V style section shrinking.

MaskRay commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1854,601 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+
+Undefined Weak Symbols
+
+An undefined weak symbol has the value 0. As the resolver function
```
This might be the glibc behavior, but musl doesn't have the special `__dl_tlsdesc_undefweak`. I think it's better to allow flexibility and require a particular behavior on undefined weak TLS.

smithp35 replied:
This part is just an example of what can be done. I've written at the top of the section
> The TLS resolver functions are not standardized by this ABI as they are internal to the dynamic linker

and

> These examples are for illustrative purposes only

I'll see if there's anything I can do to state that there is no requirement to implement a specific resolver function.

smithp35 committed:
[sysvabi64] Review comments from Claudio

* Edits to split up the bullet points in How to denote TLS
  in source.
* Changed program-own state to process-state as the thread-id
  may not be stored separately from the programs data.
* Removed typically from some of the descriptions as the typically
  will almost always be the case for a sysvabi platform.
* Linked alignment padding to the definition.
* Provided a bit more information about generation counters.

smithp35 committed:
[sysvabi64] Review comments from Maskray

* Rearranged formulas and used TCBsize to make it clearer.
* Taken out "significant" from a significant number of dynamic
  linkers.
* Give reason for using relaxation rather than optimization.
* Clarify that there is no requirement to implement any TLSDESC
  resolver given in the sysvabi.

smithp35 reviewed (commented):
Thanks very much for the review.

I've updated based on this and some comments I received internally.

smithp35 committed:
fix CI due to broken reference

smithp35 committed:
[sysvabi64] Correct local exec instruction to use tp as input

Change the input register in add xn, xn, :tprel_hi12:var, lsl #12
to the thread pointer tp. We want to calculate the offset from the
thread pointer so it needs to be an input of the add.

smithp35 committed:
[sysvabi64] document TLSDESC resolver extension register reqs

Document the decision in the GCC mailing list thread
TLSDESC clobber ABI stability/futureproofness?
https://gcc.gnu.org/legacy-ml/gcc/2018-10/msg00112.html

TLSDESC resolver functions assume that any registers added
by an extension are caller saved for a TLSDESC call.

A brief summary:

Dynamic TLS may be lazy allocated upon the first use of a TLSDESC
resolver. This may involve calls to heap allocation functions
provided by the user, which may use registers from extensions
like SVE and SME. As the resolver function can't know what is
saved it would have to save all SVE and SME state. This would
be way more expensive than a caller save, and an older libc
written prior to the introduction of the extension would be
unaware of them so the caller has to do the save.

* The SVE and SME state is already

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+The Thread Pointer ``TP`` and also the address of the start of the
+``TCB``, must satisfy the requirement.
+
+``TP ≡ 0 (modulo PT_TLS.p_align)``.
```
I originally had this written out in words ``TP congruent-to 0 (modulo PT_TLS.p_align)`` . This is more easily searched for, but it does get somewhat verbose to write multiple times.

I think I'll start with the expression for pad-size, rather than the derivation. I can also introduce the simpler max(TCBsize, PT_TLS.p_align) as well which can be understood intuitively, it is just a pity it also requires PT_TLS.p_vaddr to be 0 modulo PT_TLS.p_align.

The derivation can come later for the people that want to understand where it came from.

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+
+The code sequences below return the offset of the TLS variable from
+``tp`` in ``x0``.  To get the address of the TLS variable requires
+additional code to add ``x0`` to be added to ``tp``, this is not part
```
Yes, thanks for spotting.

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+additional code to add ``x0`` to be added to ``tp``, this is not part
+of the ABI required TLSDESC code sequence.
+
+Small Code Model
```
I'll use a semicolon. I think I'm already at the lowest level of header.

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+knows that the TLS variable is defined in the same module as the code
+that is accessing the variable. In this case the offset of the TLS
+variable from the start of the module's TLS block is a static link
+time constant. Instead of dynamically calculating the offset of the
```
ACK

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+TLS variables from the thread pointer are static link time
+constants. The code sequences are the same for all code models.
+
+The instruction sequences below are not ABI. Using the instructions
```
I've used are "not required by the ABI, but using".

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+    movz    xn, :tprel_g1:var // R_AARCH64_TLSLE_MOVW_TPREL_G1 var
+    movk    xn, :tprel_g0:var // R_AARCH64_TLSLE_MOVW_TPREL_G0_NC var
+
+TLS Descriptors
```
I'll take a look at this tomorrow.

smithp35 replied:
ACK

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+implemented.  Due to the restrictions on calling convention, the
+resolver routines must be written in assembly language.
+
+Static TLS Specialization
```
ACK

smithp35 reviewed (commented):
Thanks very much for the comments. I'll hopefully have a new patch tomorrow.

smithp35 committed:
[sysvabi64] Simple fixes from Ties' review [NFC]

smithp35 committed:
Update description of the generation counter and __tls_get_addr

Include a pseudo code description of __tls_get_addr with deferred
TLS for dynamic modules.

smithp35 committed:
Reword derivation of TLS padding to make it easier to understand

Use integers modulo m to avoid excess use of (modulo m).
Explain the congruence symbol.
Put expression first so derivation is optional.

smithp35 commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1844,6 +1857,635 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+Static link time TLS Relaxations
+--------------------------------
+
+Relaxation is a term used by the TLS literature such as ELFTLS_ to
```
ACK.

smithp35 committed:
Move around some paragraphs to introduce some topics before they
are used.

smithp35 reviewed (commented):
Thanks for the review comments. I've made the following updates:
* Simple NFC text changes.
* Better description of deferred TLS and generation count.
* Reworded the padding size derivation.
* Moved paragraphs around to make the flow a bit easier.

Should be visible as 4 separate commits

smithp35 committed:
[sysvabi64] Move TLSDESC resolvers to separate design doc

The TLSDESC resolver functions are not ABI so we can move them
out of the sysvabi64 document. Providing some examples that can
be used by a dynamic linker is still useful so move this to the
design documents section.

Add a comment about DTV surplus TLS that permits a dynamic loader
to dlopen a DSO with initial-exec TLS. There can be a small
number of performance critical shared-libraries that use initial
exec TLS, but are expected to be opened via dlopen, particularly
by scripting languages like python.

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -22,7 +23,15 @@
-.. _TLSDESC: http://www.fsfla.org/~lxoliva/writeups/TLS/paper-lk2006.pdf
+.. _ELFTLS: https://www.uclibc.org/docs/tls.pdf
+.. _TLSDESC: http://www.fsfla.org/~lxoliva/writeups/TLS/RFC-TLSDESC-ARM.txt
+.. _TLSDESCRES: https://github.com/ARM-software/abi-aa/tree/main/design-documents/tlsdesc-resolvers.txt
```
nit: file extension should probably be `.rst`

smithp35 replied:
ACK

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+*  (Most local) Automatic data (stack variables, instanced once per function
+   activation, per thread).
+
+Rules governing thread local storage on AArch64
```
This section probably should be named "Scope" and the part about `thread_local` and `__thread` should probably elsewhere.

smithp35 replied:
Will have a think to see where best to split out the source parts.

smithp35 replied:
I've decided to remove the source parts as they are out of scope, and everyone knows what they are anyway.

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+      return dtv[module_id][offset];
+    }
+
+The calculation in __tls_get_addr is the most general and it can be
```
nit:
```
s/__tls_get_addr/``__tls_get_addr``
```

smithp35 replied:
ACK

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+thread's DTV is updated, and the TLS for the ``module_id`` is
+allocated if it is not present.
+
+In pseudo code
```
I think this pseudo code listing is not necessary since we already have the description of the operation above. If both are retained, the code and the description may diverge over time.

smithp35 replied:
OK, I wanted to include the pseudo code in case the description wasn't good enough, but if it is then I can remove it. There are other sources to find pseudo code.

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+    }
+
+The calculation in __tls_get_addr is the most general and it can be
+applied to both static and dynamic TLS. There are four defined models
```
> There are four defined models

This probably should start new paragraph and probably a new section

smithp35 replied:
ACK

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+The calculation in __tls_get_addr is the most general and it can be
+applied to both static and dynamic TLS. There are four defined models
+of accessing TLS that trade off generality for performance. In order
+of descending generality:
```
> descending generality

nit: maybe "descending" is better?

smithp35 replied:
Will have a think. Perhaps `In descending order of generality:`

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+   4. Local Exec, can be used in the executable for TLS variables
+      defined in the executables static TLS block.
+
+SystemV AArch64 TLS addressing
```
> SystemV AArch64 TLS addressing

The title of this section is very similar to the previous one. I think it makes it a bit difficult to navigate this chapter.

smithp35 replied:
Will have a think to see if I can find a better one.

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+  only the descriptor dialect as this is the default dialect for GCC
+  and the only dialect supported by clang.
+
+* The thread pointer (TP) is always accessible via the ``TPIDR_EL0``
```
> The thread pointer (TP)

nit:

```
s/TP/``TP``
```

as `TP` is used like inline code below.

smithp35 replied:
ACK

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+(``PADsize``) between the TCB and the executable's TLS Block. Using
+``TCBsize`` as the size of the TCB (16 bytes), the following expression can be used to calcluate ``PADsize`` from the ``PT_TLS`` program header.
+
+``PADsize = (PT_TLS.p_vaddr - TCBsize) mod PT_TLS.p_align``.
```
style: add a `.. code-block::` instead

yury-khrustalev replied:
Same for other formulae in this section. I think it will make reading easier.

smithp35 replied:
ACK

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+resolver function.
+
+The static relocations with a prefix of ``R_AARCH64_TLSDESC_``
+targeting TLS symbol ``var``, instruct the static linker to create a
```
nit: superfluous comma after `var`?

smithp35 replied:
ACK

yury-khrustalev commented on sysvabi64/sysvabi64.rst:
```diff
@@ -1890,6 +1906,577 @@ See `MemTagABIELF64`_ and `PAuthABIELF64`_ for details of reserved tags.
+
+.. code-block:: asm
+
+    adrp xn, :gottprel: var            // R_AARCH64_TLSIE_ADR_GOTTPREL_PAGE21 var
```
nit: superfluous space before var

smithp35 replied:
ACK

yury-khrustalev reviewed (commented):
I've added a few minor comments. Otherwise this LGTM and should be approved and merged after the small fixes as per the comments. I think this chapter is very useful and provides good level of detail suitable for this ABI documentation.

smithp35 reviewed (commented):
Thanks for the comments, I'll prepare a new patch, most likely going to be Wednesday.

smithp35 committed:
[sysvabi64] Yury's review comments

* Added `` `` to some variables.
* Added some more section headings.
* Used code-blocks for formula.
* Fixed reference to design document.

smithp35 committed:
[sysvabi64] Fixup broken link [NFC]

Previous review comments changed name of a section.

yury-khrustalev reviewed (approved):
LGTM
